File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        // dp[i] = min count of coins you can have for ith amount
        int[] dp = new int[amount + 1];
        for (int i = 0; i < dp.length; i++) {
            dp[i] = amount + 1;
        }

        dp[0] = 0;

        for (int i = 1; i < dp.length; i++) {
            int curr_amt = dp[i];
            for (int coin : coins) {
                // found a solution?
                if (i - coin >= 0) {
                    // curr_amt - coin <- amt left over after taking current coin
                    dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
                }
            }
        }

        if (dp[amount] != amount + 1) {
            return dp[amount];
        }
        
        return -1;
    }
}

File: 322-coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # initialize with value bigger than possible 
        # max # of coins used is amount + 1
        dp = [amount + 1] * (amount + 1)
        # base case -> takes 0 coins to get 0 money
        dp[0] = 0

        # iterate through every target amount
        for i in range(1, amount + 1):
            # check every coin
            for c in coins:
                if i - c >= 0:
                    # 1 (current coin) + cost of the remainder (a - c)
                    dp[i] = min(dp[i], 1 + dp[i - c])
        
        return dp[amount] if dp[amount] != amount + 1 else -1
